onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/framework/data_types.h"
+#include "core/framework/sparse_tensor.h"
+
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+
+SparseTensor::SparseTensor(MLDataType elt_type,
+                           const TensorShape& shape,
+                           size_t nnz,
+                           void* values_data,
+                           void* indices_data,
+                           const OrtAllocatorInfo& allocator_info)
+    : values_(elt_type, TensorShape({static_cast<int64_t>(nnz)}), values_data, allocator_info, 0),
+      indices_(DataTypeImpl::GetType<int64_t>(),
+               TensorShape({static_cast<int64_t>(nnz), static_cast<int64_t>(shape.NumDimensions())}),
+               indices_data, allocator_info, 0),
+      shape_(shape) {}
+
+SparseTensor::SparseTensor(MLDataType elt_type,
+                           const TensorShape& shape,
+                           size_t nnz,
+                           std::shared_ptr<IAllocator> allocator)
+    : values_(elt_type, TensorShape({static_cast<int64_t>(nnz)}), allocator, 0),
+      indices_(DataTypeImpl::GetType<int64_t>(),
```

Comment:
This Tensor ctor is documented as being deprecated. Based on that, do we need this type of ctor for SparseTensor?

Reply:
I think that this is worth discussing. I saw the deprecation comment. Yet, our execution-frame/ml-value/datatypes implementation that collectively manage allocation/deallocation still use it. We could clean that up to avoid using the deprecated Tensor constructor. However, end-users may still find the deprecated constructor useful: the difference is who owns and takes responsibility for the memory. A user of ORT inference may find it convenient to delegate responsibility to the Tensor implementation (and, similarly, the SparseTensor implementation). Meanwhile, I am using the same strategy for allocation/deallocation in execution-frame/ml-value/datatypes as the existing one, so this constructor is used. (We could do this cleanup, if desired, as a separate PR, I feel.)

Reply:
@snnn I believe you marked this Tensor ctor as deprecated in 8e0fff7b. If we still use it internally, should it be marked as deprecated?

e.g.
execution_frame.cc
```
  //no memory pattern, or the pattern is not correct.
  std::unique_ptr<Tensor> p_tensor = std::make_unique<Tensor>(element_type, shape, alloc);
```